src/core/server/saved_objects/migrationsv2/index.ts

```diff
@@ -0,0 +1,240 @@
+    return stateP;
+  } else {
+    return throwBadControlState(control);
+  }
+};
+
+export const next: NextAction = (client: ElasticsearchClient, state: State) => {
+  // If for every state there is only one action (deterministic) we can define
+  // next as a state -> action set.
+  const delay = <F extends () => any>(fn: F): (() => ReturnType<F>) => {
+    return () => {
+      return state.retryDelay > 0
+        ? new Promise((resolve) => setTimeout(resolve, state.retryDelay)).then(fn)
+        : fn();
+    };
+  };
+  const control = state.controlState;
+  if (control === 'INIT') {
+    return delay(Actions.fetchAliases(client, ['.kibana', '.kibana_current', '.kibana_7.11.0']));
+  } else if (control === 'SET_SOURCE_WRITE_BLOCK') {
+    return delay(Actions.setIndexWriteBlock(client, state.source));
+  } else if (control === 'CLONE_SOURCE') {
+    return delay(Actions.cloneIndex(client, state.source, state.target));
+  } else if (control === 'FATAL' || control === 'INIT_NEW_INDICES') {
+    return null;
+  } else if (control === 'DONE') {
+    return null; // DONE
+  } else {
+    return throwBadControlState(control);
+  }
+};
+
+/**
+ * A state-action machine for performing Saved Object Migrations.
+ *
+ * Based on https://www.microsoft.com/en-us/research/uploads/prod/2016/12/Computation-and-State-Machines.pdf
+ *
+ * The state-action machine defines it's behaviour in steps. Each step is a
+ * transition from a state s_i to the state s_i+1 caused by an action a_i.
+ *
+ * s_i   -> a_i -> s_i+1
+ * s_i+1 -> a_i+1 -> s_i+2
+ *
+ * Given a state s1, `next(s1)` returns the next action to execute. Actions are
+ * asynchronous, once the action resolves, we can use the action response to
+ * determine the next state to transition to as defined by the function
+ * `model(state, response)`.
+ *
+ * We can then loosely define a step as:
+ * s_i+1 = model(s_i, await next(s_i)())
+ *
+ * When there are no more actions returned by `next` the state-action machine
+ * terminates.
+ */
+export async function migrationStateMachine(client: ElasticsearchClient, kibanaVersion: string) {
+  let state: State = {
+    controlState: 'INIT',
+    kibanaVersion,
+    aliases: {},
+    source: '',
+    target: '',
+    error: undefined,
+    retryCount: 0,
+    retryDelay: 0,
+    log: [],
+  };
+
+  let controlStateStepCounter = 0;
+
+  console.log(chalk.magentaBright('INIT\n'), state);
+  let actionThunk: ActionThunk | null = next(client, state);
+
+  while (actionThunk != null) {
+    const actionResult = await actionThunk!().catch((error) => ({ error }));
+    if (actionResult == null) {
+      throw new Error('Action must return a result.');
+    }
+    console.log(chalk.cyanBright('ACTION RESPONSE\n'), actionResult);
+
+    const newState = model(state, actionResult);
+    if (newState.log.length > 0) {
+      newState.log.forEach((log) => console[log.level](log.message));
```

Comment:
You don't want to use the platform logging system to be able to run migration from CLI? We can use a `ToolingLog`-compatible interface.